### ISSUE TITLE
comp_pcc returns 1.0 if one input is filled with all the same value

### ISSUE THREAD

**Issue by jasondavies** [tenstorrent/tt-metal#24928]
### Describe the bug

`assert_with_pcc` incorrectly passes if one tensor is full of a single value.

The root cause is that `comp_pcc` takes `np.min` of the correlation matrix.  When one input has a standard deviation of zero, this results in a matrix of `[[--, --], [--, 1.0]]`, whose minimum is 1.0.

One approach could be to pick corr_matrix[0, 1], but there is also a check in the same function for masked values and it returns 1.0 in that case (not sure why).

### Steps to reproduce the issue

```py
import torch

from tests.ttnn.utils_for_testing import assert_with_pcc

def test_pcc():
    shape = (100,)
    assert_with_pcc(torch.full(shape, 1.0), torch.rand(shape), 1.0)
```

### Expected behavior

The test should fail.

**Comment by tt-aho** [comment]
We did address this a long time ago, but it seems like we had multiple copies of the comp_pcc function and not all of them were updated, and the main copy we ended up using wasn't the fixed version.
Ex the changes here were meant to address this issue with constant tensors (this also changed the return for mask values to 0, but that was reverted back to 1. I do not recall the exact reasoning for this at the moment).
https://github.com/tenstorrent/tt-metal/commit/ef6fc09c95c5992d1681ac1f65265775d0a07a22

**Comment by fahmedchanTT** [comment]
@tt-aho Assigning to you to work through propagating the fixes

**Comment by tt-aho** [comment]
@ayerofieiev-tt is there someone from ttnn who can take over this task?
I have a draft PR unifying all of the comp_pcc functions here #25087.
When one tensor is constant we now log a warning and run a default allclose check instead now, however there are a lot of test fails in CI with this change since previously PCC would always be incorrectly reported as 1 when there were constant tensors.

https://github.com/tenstorrent/tt-metal/actions/runs/16295787279

The work here would be to just go through the failures, and most likely switch to a slightly looser allclose check (comp_allclose with adjusted atol/rtol vs comp_pcc). They should also run most of the functional pipelines, not just APC.

**Comment by ayerofieiev-tt** [comment]
@cmaryanTT , can the team take over this one please?

**Comment by cmaryanTT** [comment]
We're going to use this as an impetus to fix tests that over-rely on PCC, and should have been using equality or allclose type checks right from the start.

Taking a greedy approach, we'll start with Eltwise (there seems to be others in split and softmax). Assigning to @KalaivaniMCW to find an owner.

**Comment by nmauriceTT** [comment]
I am not a huge of fan of PCC as an 'accuracy' metric for OPs (things might be different for models). It measures correlation, not equality.
And it's very easy to build tests that succeed even though the output is completely wrong.

For instance: 

<img width="250" height="250" alt="Image" src="https://github.com/user-attachments/assets/d27d41a2-02a5-4887-8170-327ada763342" />
<img width="250" height="250" alt="Image" src="https://github.com/user-attachments/assets/a4c9f15d-95ad-40c2-aec5-fae299b4a97b" />
<img width="250" height="250" alt="Image" src="https://github.com/user-attachments/assets/fbbaa861-f863-4cb8-85a4-f22ab5ff7dc5" />

Regarding this specific issue, there are two problems. One is 'theorical' **(1)**, the other has to do with the current implementation **(2)**. 

**(1)** PCC is defined as follows:

<img width="320" height="50" alt="Image" src="https://github.com/user-attachments/assets/a1c6dc07-72c2-4598-aff0-b8f6d64bc61f" />

If `X` is constant then its standard deviation is constant => PCC is undefined. 

Of course, calling `np.corrcoeff` returns a correlation matrix rather than a single value. That's because it computes the following when comparing two series:

<img width="260" height="50" alt="Image" src="https://github.com/user-attachments/assets/c03446f7-dd0b-4767-8068-bf6fca17c8f1" />

Notes: 
- for non constant series, diagonal *should* be 1. 
- the matrix *should* be symmetrical because PCC(X, Y) = PCC(Y, X)

**(2)** That's where the implementation issue comes: we have a 2x2 symmetrical matrix with 1.0 as diagonal; we don't need to call `np.min` on the matrix.  Taking the value at  `(0,1)` or `(1, 0)` *should* be enough, and wouldn't give 1.0 if one series is constant.

Regarding the fix: I think the 'simplest' way to address this is indeed to call `allclose()` if PCC is `NaN`.



But ideally, we should really switch to something like `comp_allclose_and_pcc` or just `assert_allclose()` for most OPs; and switch to `assert_with_ulp` for *high accuracy* ops (or even assert equal for tensor manipulation OPs that should have exact results).

**Comment by fahmedchanTT** [comment]
@nmauriceTT or @cmaryanTT Looking for an update on this issue

**Comment by cmaryanTT** [comment]
No progress, still in the eltwise backlog. There is a proposal to revamp how we do testing.

**Comment by fahmedchanTT** [comment]
@cmaryanTT - Do we have a parent issue for moving the tests over from @nmauriceTT 's suggestions ?

**Comment by VirdhatchaniKN** [comment]
Update : 
All fixes have been made and merged to main
<google-sheets-html-origin>
Before | Now, On merge
-- | --
<img width="837" height="794" alt="Image" src="https://github.com/user-attachments/assets/02d711dc-562a-4b81-acb1-f580b1aa9b85" /> | <img width="974" height="772" alt="Image" src="https://github.com/user-attachments/assets/fd12c083-85e2-4daf-bae5-5c87ee964b63" />
